lib/mongodb.js

```diff
@@ -1881,14 +1881,25 @@ MongoDB.prototype.isObjectIDProperty = function(model, prop, value, options) {
   ) {
     return true;
   } else if ('string' === typeof value) {
+    // values doesn't look like an ObjectID
+    if (!value.match(/^[0-9a-fA-F]{24}$/)) return false;
+
+    // value looks like an ObjectID, and it should be treated as one
+    // because `prop.mongodb.dataType` is set to 'ObjectID'
+    // this condition overrides the `strictObjectIDCoercion` setting
+    if (prop.id && prop.mongodb && prop.mongodb.dataType === 'ObjectID') return true;
```

Comment:
Why is it important to check if `prop.id` is set? When `prop.type` is set to `ObjectID`, we consider the property as ObjectID-type regardless of `prop.id` value.

Can we do the same for `{mongodb: {dataType: 'ObjectID'}}`?

---

Also, should we support `dataType` set to ObjectID class constructor (as opposed to string name)? Should we support different casing of string values, e.g. `objectId`, `objectid`, etc.?

Reply:
> Why is it important to check if prop.id is set?

Oh, something from the previous implementation.

Reply:
@bajtos 

1. Currently `isObjectIDProperty` is called only for id properties. Should we apply it to all the properties? 
2. Currently `strictObjectIDCoercion` is applicable only to id properties. Should it be applicable to all the properties? 
3. There are no tests asserting the behavior of properties that are defined as ObjectID, but are assigned mismatching value. The current behavior is to try to convert to the type and store the result. Eg: for model property `xid: { type: 'number' }`, `xid: 'a'` will be stored as `'xid': NaN`. How do we respond to mismatched types? Throw?

I am in favor of supporting `dataType` set to `ObjectID` class constructor (as opposed to string name). This lays the foundation of supporting `ObjectID` at the higher levels eventually.

I think all of the above are desirable refactoring we should do, although it increase the scope of this PR.